core/src/main/scala/akka/kafka/KafkaConsumerActor.scala

```diff
@@ -190,7 +190,8 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
             }
         }
         //check the we got only requested partitions and did not drop any messages
-        require((rawResult.partitions().asScala -- requests.keys).isEmpty)
+        require((rawResult.partitions().asScala -- partitionsToFetch).isEmpty,
+          s"Unexpected records polled. Expected: $partitionsToFetch, result: ${rawResult.partitions()}, consumer assignment: ${consumer.assignment()}")
```

Comment:
since `poll` is performance critical we should not use `require`. It allocates a function for the second parameter. please use plain `if (...) throw IllegalArgumentException`